src/index/sparse/sparse_inverted_index.h

```diff
@@ -0,0 +1,457 @@
+        }
+        return Status::success;
+    }
+
+    Status
+    Load(MemoryIOReader& reader) {
+        std::unique_lock<std::shared_mutex> lock(mu_);
+        int64_t rows;
+        readBinaryPOD(reader, rows);
+        use_wand_ = rows > 0;
+        rows = std::abs(rows);
+        size_t dim;
+        readBinaryPOD(reader, dim);
+        readBinaryPOD(reader, value_threshold_);
+
+        raw_data_.resize(rows);
+        inverted_lut_.resize(dim);
+        if (use_wand_) {
+            max_in_dim_.resize(dim);
+        }
+
+        for (int64_t i = 0; i < rows; ++i) {
+            auto& row = raw_data_[i];
+            size_t count;
+            readBinaryPOD(reader, count);
+            row.reserve(count);
+            reader.read(row.data_ptr(), count * (sizeof(table_t) + sizeof(T)));
+            add_row_to_index(row, i);
+        }
+
+        return Status::success;
+    }
+
+    // Non zero drop ratio is only supported for static index, i.e. data should
+    // include all rows that'll be added to the index.
+    Status
+    Train(const SparseRow<T>* data, size_t rows, float drop_ratio_build) {
+        if (drop_ratio_build == 0.0f) {
+            return Status::success;
+        }
+
+        std::vector<T> vals;
+        // TODO: maybe i += 10 to down sample to speed up.
+        for (size_t i = 0; i < rows; ++i) {
+            for (size_t j = 0; j < data[i].size(); ++j) {
+                vals.push_back(fabs(data[i][j].val));
+            }
+        }
+        auto pos = vals.begin() + static_cast<size_t>(drop_ratio_build * vals.size());
+        std::nth_element(vals.begin(), pos, vals.end());
+
+        std::unique_lock<std::shared_mutex> lock(mu_);
+        value_threshold_ = *pos;
+        drop_during_build_ = true;
+        return Status::success;
+    }
+
+    Status
+    Add(const SparseRow<T>* data, size_t rows, int64_t dim) {
+        std::unique_lock<std::shared_mutex> lock(mu_);
+        auto current_rows = n_rows_internal();
+        if (current_rows > 0 && drop_during_build_) {
+            LOG_KNOWHERE_ERROR_ << "Not allowed to add data to a built index with drop_ratio_build > 0.";
+            return Status::invalid_args;
+        }
+
+        if (inverted_lut_.size() < (size_t)dim) {
+            inverted_lut_.resize(dim);
+            if (use_wand_) {
+                max_in_dim_.resize(dim);
+            }
+        }
+
+        raw_data_.reserve(current_rows + rows);
+        for (size_t i = 0; i < rows; ++i) {
+            raw_data_.emplace_back();
+            raw_data_[current_rows + i].copy_from(data[i]);
+            add_row_to_index(data[i], current_rows + i);
+        }
+        return Status::success;
+    }
+
+    void
+    Search(const SparseRow<T>& query, size_t k, float drop_ratio_search, float* distances, label_t* labels,
+           size_t refine_factor, const BitsetView& bitset) const {
+        std::shared_lock<std::shared_mutex> lock(mu_);
```

Comment:
nit: you could move the lock of `mu_` a little bit down further, because in the following lines you do nothing but `std::fill` and not using any of the fields of this struct

Reply:
Done